Mapblock mesh LOD

Handmade9485 opened the pull request:
This PR draft adds basic LOD mesh generation.

It works by changing how a mapblock mesh is generated depending on the distance to the player. Only certain node types are included in LOD generation, so for example plantlikes are not rendered. The allowed nodes are then used for a greedy mesher to further reduce vertex counts. The less detailed the LOD, the fewer node samples are used for mesh generation.

After a certain threshold, the mesh is rendered with only one material, without a texture to reduce material swaps.

The threshold after which LODs should be used, the rate at which LOD level increases, and the threshold for textureless mesh generation are configurable in the settings.

The LOD mesh generation is significantly faster than regular mesh generation, on less detailed meshes. The meshes also use considerably less RAM, since their vertex count is a lot smaller. The LODs also yield a performance improvement in my testing.

I redid this benchmark now, with 5.15 merged.

These stats were gathered by standing still in a pre-emerged world with 4000 node render and send distance, and a mesh chunk size of 16, for 15 minutes. They show how these system resources change over time or as more chunks get generated. The lod settings had a threshold of 15 blocks, a quality of 2, and a texture threshold of 2.

<img width="1815" height="955" alt="image" src="https://github.com/user-attachments/assets/c917ee2e-66c1-49d8-bdf1-67191996b4cd" />

<img width="1920" height="1080" alt="screenshot_20260214_143113" src="https://github.com/user-attachments/assets/e228334e-5b24-4bcb-af5a-d8837fdd209c" />

## To do

This PR is a Work in Progress

- [x] Fix textureless mesh day/night cycle. The materials used for these meshes currently dont get lighter/darker depending on the time of day.
- [ ] Add a way for games to opt into supporting LODs.
- [x] Clean up the extremely hacky parts or the code.

### To do maybe

- [ ] ~~Add a new kind of vertex. Since the LODs are always exactly grid aligned and are always cuboid, they wouldnt need 9 f32 fields for their position, size, and normal per vertex. The textureless meshes also wouldnt need any uv information.~~
- [ ] ~~Pre-generate multiple LODs at once, to easily swap between them when a player moves closer to or further away from  a certain chunk or uses the zoom function.~~
- [ ] ~~Persist the meshes client side, so they can be loaded immediately upon loading a world. (Might be impossible due to mesh chunk size possibly changing or when playing on servers.)~~

## How to test

These are not very detailed instructions since there are so many different combinations of nodes and settings.

1. Set block send distance to a high value.
2. Try different LOD settings
3. Try different mesh chunk sizes, minimum vertices and other performance settings
4. Try different mods and games
5. Observe mesh gen over and underground and at different times of day

Commit:
lod tests

Commit:
trying lod stuff

Commit:
trying lod stuff

Commit:
trying lod stuff

Commit:
Merge branch 'master' of https://github.com/minetest/minetest

Commit:
lod experiments

Commit:
corrected merge conflicts

Commit:
added bandaid lods

Commit:
exponential size scaling for lods reduced to linear

Commit:
added client side lod meshes

Commit:
fixed lod lighting

Commit:
merged luanti 5.11.0

Commit:
fixed edge cases

Commit:
improved lod performance through refactoring

Commit:
fixed lod lighting in caves

Commit:
fixed lod UVs; changed the default settings

Commit:
re enabled lod multithreading

Commit:
added blocky lods

Commit:
some code clean up

Commit:
fixed holes in lods

Commit:
fixed lod lighting

Commit:
fixed lod uvs

Commit:
added lod slanted setting

Commit:
re added leaves to render on slantes lods

Commit:
1x1x1 lods are shown again to avoid holes in the world

Commit:
removed duplicate multithreading from lods

Commit:
correctly centered lods on current pos if using larger mesh sizes

Commit:
minor lod fixes

Commit:
fixed water rendering as wrong kind of lod

Commit:
skipped rendering small quads on larger lods

Commit:
changed slanted threshold and fixed skipping of still large enough lods

Commit:
Fixed simple leaves not showing on lods

Commit:
Added minimum lod setting

Commit:
Added minimum lod setting

Commit:
Changed LOD defaults

Commit:
Changed near LOD calcs to facilitate future greedy meshing

Commit:
Fixed lod overflow issues

Commit:
Some cleanup and removed traces of materialless branch

Commit:
More LOD cleanup and performance improvement for close up water

lhofhansl committed:
Monoblocks: optimize blocks that contain a single type of node

lhofhansl committed:
Check for mono-block after copyTo

lhofhansl committed:
Make sure monoblocks account for all node params

lhofhansl committed:
Review comments

lhofhansl committed:
speed up getNode memory access

lhofhansl committed:
getBlockNodeIdMapping signature

lhofhansl committed:
track memory correctly

lhofhansl committed:
style

lhofhansl committed:
loose ends and a unit test

lhofhansl committed:
slight improvement

Commit:
Added greedy meshing to lods

lhofhansl committed:
address review

lhofhansl committed:
Use unique_ptr and std::copy

lhofhansl committed:
allow copy to vmanip directly from a mono-block

lhofhansl committed:
hoist is_mono_block out of serialization loop

lhofhansl committed:
Start every block as monoblock, and set m_is_mono_block in reallocate

lhofhansl committed:
minor cleanups

lhofhansl committed:
test fix/improvement

lhofhansl committed:
alignment

lhofhansl committed:
review comments

lhofhansl committed:
start blocks out a expanded block (not monoblocks)

lhofhansl committed:
slight improvements, shrink also when deserializing from network

Commit:
more greedy meshing attempts

Commit:
greedy mesher uses u64 instead of bitsets now

Commit:
attempted but failed lighting on greedy meshes

Commit:
Merge tag '5.12.0' of https://github.com/minetest/minetest into convex_hull

5.12.0

Commit:
fixed overflow on mesh grid size 16

Commit:
merge monoblocks

Commit:
fixed overflow on mesh grid size 16 part 2

Commit:
added lod resolution

Commit:
increased view range to 8k

Commit:
moved greedy lod generation to its own function

Commit:
moved greedy lod generation to its own function yet again

Commit:
greedy mesher only draws most common node types

Commit:
used u64 instead of std::bitset
called by reference

Commit:
further progress using u64 instead of std::bitset

Commit:
finished switching to u64 for bitsets

Commit:
i dont remember what i did here

Commit:
lod generation simplified

Commit:
fixed lod lighting

Commit:
fixed sides of lod not rendering

Commit:
more lod sampling and profiling

Commit:
actually added profiling

Commit:
excluded padding node in lod generation, only generate minimap for non lod,

Commit:
Merge tag '5.14.0' of https://github.com/minetest/minetest into convex_hull

5.14.0

# Conflicts:
#	shell.nix
#	src/client/camera.cpp
#	src/mapblock.cpp
#	src/mapblock.h
#	src/unittest/test_mapblock.cpp

Commit:
flattened some datastructures

Commit:
fixed segfault due to getNodeUnsafe, simplified some multiplications and map access

Commit:
calc decimal places on mesh gen duration

Commit:
more minor lod improvements

Commit:
removed slanted lods

Commit:
started mono material lods

Commit:
mono material lod depend on node color

Commit:
removed space formatting

Commit:
added lod mono material setting

Commit:
minor clean ups and naming conventions

Commit:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull
Merged upstream into feature

sfan5 commented:
Nice!

I can see a few pixels of sky leaking through in the screenshot. Is that due to the reduction code arriving at different terrain heights in neighboring blocks?

>   Add a way for games to opt into supporting LODs.

Do you foresee a reason to opt out?

> Add a new kind of vertex. [...]

You can try, but you will have to mess with the legacy `COpenGLDriver`. Or simply declare that unsupported and only allow LODs on the modern driver.

> Persist the meshes client side

I doubt this is going to be worth it. Invalidating the cache is going to be extremely annoying.

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -1819,3 +1822,300 @@ void MapblockMeshGenerator::generate()
+	std::bitset<NodeDrawType_END> solid_set;
+	solid_set.set(NDT_NORMAL);
+	solid_set.set(NDT_NODEBOX);
+	solid_set.set(NDT_ALLFACES);
```
so those outside this and the other set don't get rendered at all?
I expect that to be problematic, since you can have:
* very big plantlike nodes
* meshnodes with arbitrary contents
* nodeboxes or anything else really

the lack of these would be immediately obvious in lod meshes

Handmade9485 replied:
you can have:
> 
> * very big plantlike nodes
> * meshnodes with arbitrary contents
> * nodeboxes or anything else really
> 
> the lack of these would be immediately obvious in lod meshes

the plantlikes looked horrible when i tested that. Sure, itt would be obvious if someone built a megastructure from them and it just disappeared, but it also looked horrible to have 200ft daffodil rising from some distant hill.

nodeboxes should just display as full nodes. their textures might get a little mangled. i arbitrarily included them for now, but they might be worth some more thought if actual problems arise

meshnodes would be a bit of a headache if we dont want to force devs to package lod models themselves. honestly have no better solution for them than this

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -1819,3 +1822,300 @@ void MapblockMeshGenerator::generate()
+						irr_vertices[2] = video::S3DVertex(vertices[2], s_normals[direction], color, uvs[2]);
+						irr_vertices[3] = video::S3DVertex(vertices[1], s_normals[direction], color, uvs[3]);
+					}
+					m_collector->append(m_is_mono_mat ? s_static_tile : tile, irr_vertices, 4, quad_indices, 6);
```
wait what exactly does the LOD mesher do (compared to the normal one) if `!m_is_mono_mat`?

sfan5 commented on src/client/mesh_generator_thread.cpp:
```diff
@@ -243,8 +244,20 @@ void MeshUpdateWorkerThread::doUpdate()
+		// /*
+		//  * Calculate LOD
+		//  */
+		const v3s16 cam_pos = floatToInt(m_client->getCamera()->getPosition(), BS) / MAP_BLOCKSIZE // current player block
```
`m_client->getCamera()` is probably not safe in this thread

Handmade9485 replied:
ive looked a bit more into this and cant find another way. I personally didnt experience any visibly undefined behavior from this, but also dont wanna risk it either.
So what exactly do you mean by "not safe"? Can the camera pointer just change during runtime?

Handmade9485 commented:
> I can see a few pixels of sky leaking through in the screenshot. Is that due to the reduction code arriving at different terrain heights in neighboring blocks?

Yes, some of those are the LOD boundaries. Im honestly not sure what can be done about that, since i wouldnt know hoe to identify those boundaries in code :/
Also, the black artifacts are just other gaps from neighboring blocks not being present in the map due to send optimization.

> Do you foresee a reason to opt out?

Yes. Some games might want some more fine grained control. Maybe there could be some scenarios of tiny floating islands where its important to see some detail. Maybe like a puzzle game where you have to look at distant objects for clues.

Handmade9485 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -588,7 +588,15 @@ void PartialMeshBuffer::draw(video::IVideoDriver *driver) const
+	mat.FogEnable = true;
+	mat.MaterialType = video::EMT_SOLID;
+	return mat;
+}();
```
yeah, this is one of the hacky parts. How would one create such a shader material?

sfan5 replied:
check the `getShader` calls in nodedef.cpp

nauta-turbidus replied:
Well, maybe there is a place for textureless mesh too... but feels like you should go textureless at such a scale that the cube loses its shape anyway...?

Handmade9485 replied:
Ok, i have some cold, hard data now. Here is a scene using a point cloud. The transition from regular mesh to point cloud is at a distance where you wouldnt be able to tell the textured and textureless meshes apart without zooming in. The point cloud however is immediately obvious.

<img width="1920" height="990" alt="screenshot_20251026_012615" src="https://github.com/user-attachments/assets/b100fa6d-9f05-46f9-b46d-7fc7b5691d77" />
<img width="1920" height="990" alt="screenshot_20251026_014319" src="https://github.com/user-attachments/assets/0793511a-31a9-46c5-874c-8b1b4fbb680d" />

The point cloud might be useful at a further distance, but i cant figure out how to pass a different size for each vertex to the shader and i wanna avoid feature creep for now.

Handmade9485 replied:
>you may be transitioning to textureless too close I think, feels like that should be another mipmap level.

textureless has the advantage, that it needs no material swaps, as opposed to mipmaps. Also, if you look at the border between lod and regular nodes, you cant make out the texture on the regular ones. Sure, we can talk about what default settings would make the most sense, but my main point was just that point clouds arent always equivalent to textureless meshes.

> Also, well, you probably should stuff the scale in alpha channel of vertex color?

huh, that should work. It feels a bit dirty, but theres probably no use for transparency for those points anyway.

> The pointcloud should be denser then

you mean the distant ones? Yeah, I just wrote something quick and dirty that still uses the greedy mesher under the hood, which ofc is the wrong choice for a point cloud. They just cant be too dense or wed slow down mesh generation too much.

nauta-turbidus commented:
How many LOD levels are there exactly? I feel like the "textureless" level, whatever it is, should be just a point cloud.

Handmade9485 commented:
> How many LOD levels are there exactly? I feel like the "textureless" level, whatever it is, should be just a point cloud.

its clamped to the chunk width. so while higher levels are technically possible, any visual change stops around the ballpark of level log2(16) to log2(256).

Mind explaining the point cloud a bit more? Are they already supported by irrlicht, or by luanti? Would they not need a draw call per chunk? Could all the point clouds easily be merged into one big cloud, to use only one draw call for all of them, without incurring the merging overhead of the current mesh_buffer_min_vertices for meshes?
I assume "mesh" gen would be pretty fast for them.

Unfortunately, im also not familiar enough with luanti, irrlicht, or rendering in general, to know how / where to implement such a functionality without some hand holding.

nauta-turbidus commented:
> > How many LOD levels are there exactly? I feel like the "textureless" level, whatever it is, should be just a point cloud.
> 
> its clamped to the chunk width. so while higher levels are technically possible, any visual change stops around the ballpark of level log2(16) to log2(256).
> 
> Mind explaining the point cloud a bit more? Are they already supported by irrlicht, or by luanti? Would they not need a draw call per chunk? Could all the point clouds easily be merged into one big cloud, to use only one draw call for all of them, without incurring the merging overhead of the current mesh_buffer_min_vertices for meshes? I assume "mesh" gen would be pretty fast for them.
> 
> Unfortunately, im also not familiar enough with luanti, irrlicht, or rendering in general, to know how / where to implement such a functionality without some hand holding.

Well, you'd need to replace the `glDraw...(GL_TRIANGLES...)` call with `glDraw...(GL_POINTS...)`. That way you can go to point per node, then point per x nodes, point per mapblock, even point per mapchunk. At that point you're limited by data sent from the server only.

Edit: Since a point cloud is just an array of colored points (in general), you can easily merge as many mapblocks/mapchunks into a single draw call. Even accross multiple "point" LODs (as above, point per node, point per X nodes, point per mapblock, point per mapchunk even), points are probably the same and you probably can pack them all into a single large draw call.

Commit:
Moved lod mesh gen to its own file

Commit:
removed merge error

Handmade9485 commented:
> Also did you check if this actually effective? We removed our greedy mesher a while ago since we couldn't identify a performance difference and it caused some problems.

Ok, i did some tests now. The first one i did was with mesh chunk 4 and fairly low lod settings. The difference between greedy and regular meshes was pretty small in this test.
<img width="1885" height="975" alt="image" src="https://github.com/user-attachments/assets/cb18fd13-2aa4-4dd3-9668-181201c6cea4" />


The next test was done with mesh chunk 3 and pretty high lod settings. Here there was a pretty clear difference in that the greedy meshes took longer until frame time became absolutely abysmal. The RAM gap between greedy and regular meshes also grew by a wider margin.
<img width="1885" height="975" alt="image" src="https://github.com/user-attachments/assets/f178db6d-1396-4f23-92f0-f0b77fe076f1" />


What were the issues you had with the previous greedy mesher?

Commit:
fixed textureless lighting by extending the shader/materials

sfan5 commented on irr/include/EMaterialTypes.h:
```diff
@@ -51,6 +51,8 @@ enum E_MATERIAL_TYPE
 	pack_textureBlendFunc (for 2D) or pack_textureBlendFuncSeparate (for 3D). */
 	EMT_ONETEXTURE_BLEND,
 
+	EMT_TEXTURELESS,
```
you can pass this as a shader parameter, don't add a new material.

Handmade9485 replied:
Ok, I think i got a pretty ok idea of what this entails, but just wanna confirm it before putting in too much work.

- add a new UniformSetter and implement whatever is necessary for a CachedPixelShaderSetting to pass a boolean
- add an "isTextureless" attribute the driver
- add another attribute like that to ISceneNode
- set the driver attribute in [drawAll](https://github.com/luanti-org/luanti/blob/8350fb734abf9ead60517fb2ed7fb0da8d923de7/irr/src/CSceneManager.cpp#L454) if the node is textureless

Is this how its done or am i overlooking something simpler?

Commit:
fixed incorrect indent and removed unused code

Commit:
added tile colors for use in textureless lod meshes that dont depend on the minimap color

4562448 committed:
reordered lod directions to be in line with the rest of the engine

4562448 committed:
replaced magic number with a constant

4562448 committed:
fixed some nuclear level brain farts

4562448 committed:
added some comments

4562448 committed:
moved lod calculation outside of mesh gen thread

4562448 committed:
reimplemented std::__countr_zero in software to support other compilers

sfan5 commented on src/client/lod_mapblock.cpp:
```diff
@@ -20,6 +20,55 @@
 static const auto &quad_indices = quad_indices_02;
 
+// a binary search implementation of std::countr_zero
+u8 countr_zeros(u64 n) {
```
it's smarter to delegate this to `__builtin_ctz`, because the compiler is unlikely to be able to tell wtf your code is doing and will compile all these inefficient ifs/loops.
for a fallback a simple bitwise loop suffices. the compiler might even recognize it and replace it with its own efficient implementation.

4562448 committed:
used __builtin_ctzll instead

4562448 committed:
fixed lods not being aligned with chunks

4562448 committed:
fixed typo

sfan5 commented on src/client/lod_mapblock.cpp:
```diff
@@ -0,0 +1,345 @@
+
+				bitset column = m_slices[slice_offset + u];
+				while (column) {
+					s32 v0 = __builtin_ctzll(column);
```
to be clear this probably needs to be wrapped for compilers where this builtin is not available

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

# Conflicts:
#	client/shaders/nodes_shader/opengl_fragment.glsl
#	src/client/mapblock_mesh.h

4562448 committed:
removed textureless material

4562448 committed:
always include glasslikes in lods to be more consistent

4562448 committed:
hopefully be portable. dont have all the compilers to test, so have to rely on github workflows

4562448 committed:
added mesh nodes to lods as regular nodes
this didnt look too egregious in my testing but is worth looking further into

4562448 committed:
renamed mono material to textureless to be more consistent

4562448 committed:
added a setting to enable/disable LODs altogether

4562448 committed:
replaced spaces with tabs

4562448 committed:
fixed wrong whitespace and removed unused code

4562448 committed:
fixed usage of _BitScanForward64

4562448 committed:
fixed even more wrong indents

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

4562448 committed:
fixed a whole lot of merge mistakes

4562448 committed:
even more wrong whitespace

4562448 committed:
even more wrong whitespace

4562448 committed:
even more wrong whitespace

4562448 committed:
hope and pray i finally used _BitScanForward64 correctly

4562448 committed:
incorporate tile color for textureless lods

4562448 committed:
moved ctzll to porting

4562448 committed:
fixed big where tile color gets applied twice

4562448 committed:
reworked lod mesh gen to not need separate passes for opaque and transparent nodes;
rendered meshnodes normally when not textureless, because they looked horrible sometimes

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

# Conflicts:
#	src/client/content_mapblock.h
#	src/client/mapblock_mesh.cpp
#	src/nodedef.cpp
#	src/nodedef.h

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

# Conflicts:
#	src/client/content_mapblock.h
#	src/client/mapblock_mesh.cpp
#	src/nodedef.cpp
#	src/nodedef.h

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

# Conflicts:
#	src/client/content_mapblock.h
#	src/client/mapblock_mesh.cpp
#	src/nodedef.cpp
#	src/nodedef.h

4562448 committed:
fixed #else

4562448 committed:
fixed wrong line break

4562448 committed:
used std::array for map values instead of c style arrays

4562448 committed:
fixed wrong uvs on some lod faces

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

lhofhansl commented:
Needs a rebase again. :(
Sorry @Handmade9485 .

Handmade9485 commented:
@lhofhansl no need to be sorry.
Im just a little burnt out on this at the moment, but if the past is anything to go by, ill probably find the motivation to continue working on this in a couple of months.

And even then, Id probably try to get this feature complete before trying to resolve the merge conflicts. So, If anyone has input on how to do any of these, or if they are even feasible/necessary, let me know:

1. Making LODs be opt in for games.
Im worried about existing games, which were made without LOD support. Adding this retroactively, could destroy any game that relies on some level of visual clarity/fidelity.

2. Zooming in should show more detailed blocks.
Showing the player a zoomed in version of an LOD block wouldnt be very helpful, so the engine would need to switch from e.g. LOD2 up to LOD1. Im just not sure on the best implementation for this. Or maybe is isnt even necessary for this PR to begin with, and can be postponed.

3. Reflective water makes the LOD line pretty obvious.
No idea yet, how the engine differentiates between water and other liquids, to apply the shader only to the water, but since the LODs arent affected by that shader, it makes the seams pretty noticeable.

sfan5 commented:
First we need the necessary "infrastructure" for any kind of LODs to be implemented, something like described here: https://github.com/luanti-org/luanti/issues/15172#issuecomment-3702208670

4562448 committed:
some helpful comments

4562448 committed:
changed loops to reduce cache misses

4562448 committed:
node loading optimizations

4562448 committed:
small readability improvement

4562448 committed:
small readability improvement

4562448 committed:
small readability improvement

4562448 committed:
added lod meshgen benchmark

4562448 committed:
Merge tag '5.15.0' of https://github.com/minetest/minetest into convex_hull

5.15.0

# Conflicts:
#	shell.nix
#	src/client/mapblock_mesh.cpp

4562448 committed:
just straight up used the regular tiles for all liquids, because the mono mats wouldnt know if it would have to be reflective or not

4562448 committed:
leave holes in lod mesh when adjacent node is not loaded, like in regular mesh gen

4562448 committed:
simplified lod mesh gen for reflective water
removed generateMonoMesh as its become obsolete

kromka-chleba commented:
Are distant mapblocks still simplified to cubes as in the OP's screenshot?
I believe triangulated planes would both be more efficient and look better.
Thoughts?

4562448 committed:
some restructuring

4562448 committed:
about 5% performance increase by removing memset and thus reducing cache misses

4562448 committed:
somehow removing this doesnt harm performance

4562448 committed:
started removing greedy mesher

4562448 committed:
got rid of all greedy meshing remains

Handmade9485 commented:
Well, as difficult as it was for me to come to terms with, the greedy meshing was a mistake. Using non greedy meshes does use a bit more ram, but the mesh generation got faster by at least an order of magnitude. It also looks better and made it way simpler to keep water surfaces the same level, even across LOD borders.

So with this done, I cant think of anything else to implement at this point. It seems like I still gotta fix a few merge conflicts, but then what? Do I just click the "Ready for review" button?

<img width="1920" height="1080" alt="screenshot_20260305_001929" src="https://github.com/user-attachments/assets/a2b04a7a-7576-45f6-b973-07d688d0a784" />

nauta-turbidus commented:
> So with this done, I cant think of anything else to implement at this point. It seems like I still gotta fix a few merge conflicts, but then what? Do I just click the "Ready for review" button?

Probably, yeah. Even if there is more stuff that could be done, it's better to leave the rest (pointclouds? etc.) for future PRs (assuming this one goes well).

4562448 committed:
Merge remote-tracking branch 'refs/remotes/minetest/master' into convex_hull

# Conflicts:
#	client/shaders/nodes_shader/opengl_fragment.glsl
#	src/client/shader.cpp
#	src/client/shader.h

4562448 committed:
ok, but now i got rid of all greedy meshing related stuff

4562448 committed:
fixed incorrect lod face rotations

4562448 committed:
fixed inaccurate node neighbor detection for lod mesh gen

4562448 committed:
lods now update when moving around

nauta-turbidus commented:
> I dont really get that. Do you mean a 10 BS hysteresis?

The "back and forth part" points to hysteresis I think. And obviously, no LOD system may work properly without hysteresis.

4562448 committed:
moved reading settings outside of loops

4562448 committed:
recalculate lod if block with default lod makes its way into mesh gen

4562448 committed:
Merge branch 'master' of https://github.com/minetest/minetest into convex_hull

# Conflicts:
#	src/client/shader.cpp
#	src/client/shader.h

Handmade9485 commented on src/client/lod_mapblock.cpp:
```diff
@@ -0,0 +1,281 @@
+	while (m_data->m_side_length % m_node_width != 0) m_node_width--;
+
+	for (m_cur_node.p.Z = m_node_width - 1; m_cur_node.p.Z < m_data->m_side_length; m_cur_node.p.Z += m_node_width)
+	for (m_cur_node.p.Y = m_node_width - 1; m_cur_node.p.Y < m_data->m_side_length; m_cur_node.p.Y += m_node_width)
```
Its been a while since i last looked at this, but im fairly certain that the nodes are stored x minor. So iterating over x last **should** reduce cache misses in theory. Seeking to the surface, on the y axis will inevitably still cause cache misses, but on dense terrain like underground, or underwater, the seeking shouldnt be necessary.
Though admittedly, i dont have any data on this, and im not sure if its even possible to gather some relatively clean data without having to write some benchmarks first :/

4562448 committed:
fixed bug where lod meshes generated top and bottom faces even when obstructed

4562448 committed:
always calculate lod in mesh gen thread, because its honestly 1000x easier to handle

4562448 committed:
something
